sycl/include/sycl/detail/kernel_launch_helper.hpp

```diff
@@ -251,20 +254,325 @@ struct KernelWrapper<
+  enum class StableKernelCacheConfig : int32_t {
+    Default = 0,
+    LargeSLM = 1,
+    LargeData = 2
+  };
+
+  // This struct is used to store kernel launch properties.
+  // std::optional is used to indicate that the property is not set.
+  // In some code paths, kernel launch properties are set multiple times
+  // for the same kernel, i.e why using std::optional to avoid overriding
+  // previously set properties.
+  struct KernelLaunchPropertiesT {
+    std::optional<ur_kernel_cache_config_t> MKernelCacheConfig;
+    std::optional<bool> MKernelIsCooperative;
+    std::optional<uint32_t> MKernelWorkGroupMemorySize;
+    std::optional<bool> MKernelUsesClusterLaunch;
+    size_t MKernelClusterDims;
+    std::array<size_t, 3> MKernelClusterSize;
+
+    KernelLaunchPropertiesT()
+        : MKernelCacheConfig(std::nullopt), MKernelIsCooperative(std::nullopt),
+          MKernelWorkGroupMemorySize(std::nullopt),
+          MKernelUsesClusterLaunch(std::nullopt), MKernelClusterDims(0),
+          MKernelClusterSize{0, 0, 0} {}
+
+    KernelLaunchPropertiesT &operator=(const KernelLaunchPropertiesT &Other) {
+      if (Other.MKernelCacheConfig)
+        MKernelCacheConfig = Other.MKernelCacheConfig;
+
+      if (Other.MKernelIsCooperative)
+        MKernelIsCooperative = Other.MKernelIsCooperative;
+
+      if (Other.MKernelWorkGroupMemorySize)
+        MKernelWorkGroupMemorySize = Other.MKernelWorkGroupMemorySize;
+
+      if (Other.MKernelUsesClusterLaunch) {
+        MKernelUsesClusterLaunch = Other.MKernelUsesClusterLaunch;
+        MKernelClusterDims = Other.MKernelClusterDims;
+        MKernelClusterSize = Other.MKernelClusterSize;
+      }
+
+      return *this;
+    }
+
+    KernelLaunchPropertiesT(const KernelLaunchPropertiesT &Other) = default;
+    KernelLaunchPropertiesT(KernelLaunchPropertiesT &&Other) = default;
+    ~KernelLaunchPropertiesT() = default;
+  };
+
+  KernelLaunchPropertiesT KLProps;
+
+  // Set default values for kernel launch properties.
+  KernelLaunchPropertyWrapper() {
+    KLProps.MKernelCacheConfig = UR_KERNEL_CACHE_CONFIG_DEFAULT;
+    KLProps.MKernelIsCooperative = false;
+    KLProps.MKernelWorkGroupMemorySize = 0;
+    KLProps.MKernelUsesClusterLaunch = false;
+    KLProps.MKernelClusterDims = 0;
+    KLProps.MKernelClusterSize.fill(0);
+  }
+
+  template <sycl::ext::oneapi::experimental::detail::UnsupportedGraphFeatures
+                FeatureT>
+  static void throwIfGraphAssociated(bool HasGraph) {
+    if (!HasGraph)
+      return;
+
+    std::string FeatureString =
+        sycl::ext::oneapi::experimental::detail::UnsupportedFeatureToString(
+            FeatureT);
+    throw sycl::exception(sycl::make_error_code(errc::invalid),
+                          "The " + FeatureString +
+                              " feature is not yet available "
+                              "for use with the SYCL Graph extension.");
+  }
+
+  static void verifyDeviceHasProgressGuarantee(
```

Comment:
This is weird as well, it feels like the code is grouped based on how someone wrote it someday, not based on what logical abstractions/interfaces should be.